File: py_server/callback.py
import debug;
import _MID;
import time;

TIMEOUT_NONE = 0.0;
TIMEOUT_SHORT = 2.0;
TIMEOUT_NORMAL = 5.0;
TIMEOUT_LONG = 10.0;

CALLBACK_RESULT_UNKNOWN = 0;
CALLBACK_RESULT_SUCCESS = 1;
CALLBACK_RESULT_TIMEOUT = 2;

class MessageCallback:

    mid = None;
    func = None;
    creation_time = 0;
    timeout_len = TIMEOUT_NONE;
    remove_after_call = False;

    def __init__(self, mid, func, timeout_len = TIMEOUT_NONE, remove_after_call = False):
        self.mid = mid;
        self.func = func;
        self.timeout_len = timeout_len;
        self.remove_after_call = remove_after_call;
        self.reset_timeout();

    def reset_timeout(self):
        self.creation_time = time.time();

def make_MID_callback(mid, callback_func, timeout_len = TIMEOUT_NONE, remove_after_call = False):
    cb = MessageCallback(mid, callback_func, timeout_len, remove_after_call);
    return cb;

File: proj.win32/src/network/message/Callback.cpp
#include "network/message/Callback.h"

#include "utility/Logger.h"
#include "network/message/MID.h"
#include "network/message/Message.h"
#include "network/sockets/Socket.h"
#include "network/sockets/SocketPoll.h"

BEGIN_NETWORK_NS
BEGIN_MSG_NS

//private

//public
CallbackPtr make_MID_callback(CMID mid, CallbackFunc callback, float timeout_len, bool remove_after_call) {
    return CallbackPtr(new MIDCallback(mid, callback, timeout_len, remove_after_call));
}

void process_message_callback(MessagePtr message) {
    message->callback_result = msg::CALLBACK_RESULT_SUCCESS;
    for (int n = 0; n < message->sock->callbacks.size(); ++n) {
        if (message->sock->callbacks[n]->mid == message->mid) {
            message->sock->callbacks[n]->func(message.get());
            if (message->sock->callbacks[n]->remove_after_call) {
                message->sock->callbacks.erase(message->sock->callbacks.begin() + n);
                --n;
            }else {
                message->sock->callbacks[n]->reset_timeout();
            }
        }
    }
}

void process_all_callbacks(sock::SocketPoll& poll) {
    time_t t;
    for (int i = 0; i < poll.get_size(); ++i) {
        sock::Socket* sock = poll.get_sock_at(i);
        if (!sock) continue;
        for (int n = 0; n < sock->callbacks.size(); ++n) {
            msg::CallbackPtr& cb = sock->callbacks[n];
            if (cb->timeout_len != msg::TIMEOUT_NONE && (time(&t) - cb->creation_time) >= cb->timeout_len) {
                log_info << "callback timeout for " << cb->mid->name;
                msg::Message m;
                m.callback_result = msg::CALLBACK_RESULT_TIMEOUT;
                m.sock = sock;
                cb->func(&m);
                if (cb->remove_after_call) {
                    sock->callbacks.erase(sock->callbacks.begin() + n);
                    --n;
                }else {
                    cb->reset_timeout();
                }
            }
        }
    }
}

END_MSG_NS
END_NETWORK_NS